src/Microsoft.TemplateEngine.Tasks/Tasks/LocalizeTemplates.cs

```diff
@@ -0,0 +1,188 @@
+        }
+
+        public void Cancel() => GetOrCreateCancellationTokenSource().Cancel();
+
+        /// <summary>
+        /// Given a <paramref name="path"/>, finds and returns all the template.json files. The search rules are executed in the following order:
+        /// <list type="bullet">
+        /// <item>If path points to a template.json file, it is directly returned.</item>
+        /// <item>If path points to a template directory, path to the "&lt;directory&gt;/.template.config/template.json" file is returned.</item>
+        /// <item>If path points to a "template.config" directory, path to the "&lt;directory&gt;/template.json" file is returned.</item>
+        /// <item>If path points to any other directory and <paramref name="searchSubdirectories"/> is <see langword="true"/>, path to all the
+        /// ".template.config/template.json" files under the given directory is returned.</item>
+        /// </list>
+        /// </summary>
+        /// <param name="path">Path to search for template.json files.</param>
+        /// <param name="searchSubdirectories">Indicates weather the subdirectories should be searched
+        /// in the case that <paramref name="path"/> points to a directory. This parameter has no effect
+        /// if <paramref name="path"/> points to a file.</param>
+        /// <returns>A path for each of the found "template.json" files.</returns>
+        private IEnumerable<string> GetTemplateJsonFiles(string path, bool searchSubdirectories)
+        {
+            if (string.IsNullOrEmpty(path))
+            {
+                yield break;
+            }
+
+            if (File.Exists(path))
+            {
+                yield return path;
+                yield break;
+            }
+
+            if (!Directory.Exists(path))
+            {
+                // This path neither points to a file nor to a directory.
+                yield break;
+            }
+
+            if (!searchSubdirectories)
+            {
+                string filePath = Path.Combine(path, ".template.config", "template.json");
+                if (File.Exists(filePath))
+                {
+                    yield return filePath;
+                }
+                else
+                {
+                    filePath = Path.Combine(path, "template.json");
+                    if (File.Exists(filePath))
+                    {
+                        yield return filePath;
+                    }
+                }
+
+                yield break;
+            }
+
+            foreach (string filePath in Directory.EnumerateFiles(path, "template.json", SearchOption.AllDirectories))
+            {
+                string? directoryName = Path.GetFileName(Path.GetDirectoryName(filePath));
+                if (directoryName == ".template.config")
+                {
+                    yield return filePath;
+                }
+            }
+        }
+
+        private CancellationTokenSource GetOrCreateCancellationTokenSource()
```

Comment:
question: why do we need this logic?
does MSBuild create class per task/class per run? Should we cancel all tasks if Cancel is called?

Reply:
`Cancel` will be called from a thread other than the one running the `Execute` method. The method ensures that two threads don't step onto each other's toes when creating the ct source.

We can instantiate it earlier to avoid this, but any instance created by MSBuild will leave it un-disposed if it is not executed. It may also be created needlessly in some cases where we quit before invoking the async methods.

Reply:
I suspect we get an instance of `LocalizeTemplates` per run. Note that `_cancellationTokenSource` field is not static.

Calling `Cancel` will cancel all the tasks initiated by this `LocalizeTemplates` instance. Do you recommend another way?